### ISSUE TITLE
Should a server ignore unrecognized query parameters OR throw a 400?

### ISSUE THREAD

**Issue by pvretano** [opengeospatial/WFS_FES#128]
Should a server ignore unrecognized query parameters (i.e. not defined in the OpenAPI document) or should it throw a 400 exception?

**Comment by aaime** [comment]
How does one tell apart an unrecognized property from an unrecognized vendor parameter, or even just a cache breaking param added to make sure the request reach the origin server (there might be other ways, but this is a common practice)?

**Comment by pvretano** [comment]
Hmm ... Let me rephrase the question a little better (I hope) ... What should the server do if it encounters a query parameter that is not defined in its OpenAPI document?  Should it throw an exception or simple ignore the parameter?  My going in throught was that the server should throw a 400 because I assume that any vendor parameters would be included in the OpenAPI document.  @aaime Good point about the cache breaking parameter.  Perhaps the server should just ignore the unrecognized parameters.  Comments?

**Comment by nmtoken** [comment]
My feeling, just like in KVP, is that the server should ignore unknown parameters.  This allows extensibility

James

-----Original Message-----

**Comment by tomkralidis** [comment]
+1 to ignore. Might be too strict to return an exception causing the client to go back to the OpenAPI def.

Are there any options to not throw an error but provide a parameter like `debug=true` (default false) or something so that a client would get back a 200 which includes some `debug` object with messages?

**Comment by jvanulde** [comment]
+1 to ignore

**Comment by JivanAmara** [comment]
+1 to ignore.
I'm doing some automatic configuration of data providers, allowing many columns to become feature properties that can be filtered on.  I think it would be non-trivial to also automatically configure the openapi document based on the content of the data providers...with little real gain for users.

**Comment by pvretano** [comment]
I am fine with the ignore approach -- heck that is what WFS has done since version 0.0.13 -- BUT I am also concerned that allowing unknown or undefined parameters might break existing OpenAPI tooling.  Unfortunately I don't have enough experience using these tools (like swagger ui) to know if this is an issue we need to worry about.  Anyone want to weight in?  I found these issues that seem to indicate that OpenAPI 3.0 does allow dynamic or free-form parameters so that is encouraging ... https://github.com/OAI/OpenAPI-Specification/issues/349, https://github.com/OAI/OpenAPI-Specification/issues/367.

**Comment by cportele** [comment]
I just had a look at the OpenAPI discussions (see Peter's links) and how this is addressed in OpenAPI 3.0.

Basically you can declare support for additional parameters by defining a parameter with `type`=`object` and `style`=`form`. The name of the parameter does not matter as the actual query parameters are the names of the object properties. Say, I define a parameter `vendor_specific_parameters` (to use the OWS Common term) with `type`=`object` and `style`=`form`, then a value of the parameter (in JSON) would be something link this: `{ "my_first_parameter": "some value", "my_other_parameter": "another value" }`.

In the URI this would be expressed as `...&my_first_parameter=some%20value&my_other_parameter=another%20value`.

See https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.1.md#parameter-object and the free-form parameter example that covers this case.

As this is explicit design in the OpenAPI spec, at least in the OpenAPI conformance class I think we should require that a server raises an exception for undeclared query parameters and explain (or have a requirement?) how to include a wildcard/free-form parameter, if a server wants to express that it will ignore unknown parameters or that it supports undocumented parameters.

**Comment by cportele** [comment]
15-April-2019 Teleconference: Based on the OpenAPI spec we should require that a server raises an exception for undeclared query parameters and explain how to include a wildcard/free-form parameter, if a server wants to express that it will ignore unknown parameters or that it supports undocumented parameters. Or to explicitly include all vendor specific parameters in the API definition.

**Comment by aaime** [comment]
Just found this in CITE tests, quite surprised by the strict behavior (everything else about the new services seems pretty loose by comparison). So one cannot, for example, make up a cache boosting parameter unless the API declares one for the purpose? Scenario, there is cache in between, the client wants to make sure to get fresh data. 

Also, while this behavior is currently declared in the Features API, it's general, one would expect it to apply to all other API based services no? Forcing an extra param in the URL to get a fresh tile it's probably one of the most common way to bypass a CDN...

**Comment by cportele** [comment]
@aaime - If you want to support undocumented parameter, you can declare this in your API definition. If not, don't. The spec states how to do it in OpenAPI: http://docs.opengeospatial.org/is/17-069r3/17-069r3.html#query_parameters.

**Comment by aaime** [comment]
@cportele thanks for the follow up. I've tried to modify the API definition accordingly, here is what I came up with:

https://gist.githubusercontent.com/aaime/50c32a9b621f7e5733b59bac585dadcd/raw/699304896f4cb0a8dbfc7d7b60929577ecb0d799/test-additional-properties.json

It may be that I did it wrong, but I also cannot see any precondition check in the CITE tests, verifying if the API allows for additional parameters:

https://github.com/opengeospatial/ets-ogcapi-features10/blob/a7eb2a19fb47fc6523c82a283f18794a2f6390a7/src/main/java/org/opengis/cite/ogcapifeatures10/collections/FeaturesErrorConditions.java#L113

Happy to open a report, if you can confirm the test suite has a bug there.

**Comment by cportele** [comment]
@aaime - Your OpenAPI definition of otherParameter looks correct to me. The CITE test is clearly too simple as it just assumes that a query parameter `unknownQueryParameter` is not valid. The test needs to inspect the API definition to determine the query parameters that have been defined for the operation.

**Comment by aaime** [comment]
Thanks, I'm opening a report then.
